### PR TITLE
Client/Server plugin and Discovery

### PR DESCRIPTION
**Pull request by sbezverk**
This PR:
 - converts nsmdp into a new NSM plugin,
 - adds DiscoverRequest handler for server side and call it from nsm client,
 - fixes several issues discovered during debugging.

Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>

### CODE REVIEW

**Comment by sbezverk** [comment]
Log from the client, it prints discovered via API call network services:
```
time="2018-07-01T17:06:36Z" level=info msg="nsm client: connection to nsm server on socket: /var/lib/networkservicemesh/nsm.ligato.io.sock succeeded."
time="2018-07-01T17:06:38Z" level=info msg="nsm client: %d NetworkServices discovered from Local NSM.1"
time="2018-07-01T17:06:38Z" level=info msg="nsm client: pod's namespace is [NS(6: 3, 4026532392)]"
time="2018-07-01T17:06:38Z" level=info msg="nsm client: pod's interfaces:"
time="2018-07-01T17:06:38Z" level=info msg="Name: lo Type: device"
time="2018-07-01T17:06:38Z" level=info msg="Name: eth0 Type: veth"
time="2018-07-01T17:06:38Z" level=info msg="nsm client: requested network services:"
time="2018-07-01T17:06:38Z" level=info msg="&{Name:network-service-1 ServiceInterface:[{Type:KERNEL_INTERFACE Metadata:name:\"interface-1\"  Preference:FIRST Parmeters:<nil> XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0} {Type:MEM_INTERFACE Metadata:name:\"interface-2\"  Preference:THIRD Parmeters:<nil> XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}]}"
time="2018-07-01T17:06:38Z" level=info msg="&{Name:network-service-2 ServiceInterface:[{Type:KERNEL_INTERFACE Metadata:name:\"interface-3\"  Preference:FIRST Parmeters:<nil> XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}]}"
time="2018-07-01T17:06:38Z" level=info msg="nsm client: initialization is completed successfully, exiting..."
```

**Comment by sbezverk** [comment]
nsm client log indicating discovery of 1 NetworkService via API
```
time="2018-07-01T17:20:21Z" level=info msg="nsm client: connection to nsm server on socket: /var/lib/networkservicemesh/nsm.ligato.io.sock succeeded."
time="2018-07-01T17:20:23Z" level=info msg="nsm client: 1 NetworkServices discovered from Local NSM."
```

**Comment by sbezverk** [comment]
@mestery @edwarnicke @fkautz  PTAL when you have time..

File: cmd/nsm-init/nsm-init.go [hunk 1]
```diff
@@ -192,22 +195,57 @@ func main() {
...
+	availablaNetworkServices, err := getNetworkServices(nsmClient)
+	if err != nil {
+		logrus.Fatalf("nsm client: failed to get a list of NetworkServices from NSM, exiting...", err)
+		os.Exit(1)
+	}
 
+	if len(availablaNetworkServices) == 0 {
+		// Since local NSM has no any NetworkServices, then there is nothing to configure for the client
+		logrus.Info("nsm client: Local NSM does not have any NetworkServices, exiting...")
+		os.Exit(0)
```
Comment (mestery) [inline cmd/nsm-init/nsm-init.go]: We'd want the client to continue running so when NetworkServices are configured it would pick them up.

File: plugins/crd/plugin_impl_crd.go [hunk 2]
```diff
@@ -120,13 +120,13 @@ func networkServiceValidation() *apiextv1beta1.CustomResourceValidation {
 								"name": {
 									Type:        "string",
 									MaxLength:   &maxLength,
-									Description: "NetworkServiceEndpoints Name",
```
Comment (mestery) [inline plugins/crd/plugin_impl_crd.go]: Pedantic nit: This change is unrelated to the work being done here. I'm fine with cleanups like this as part of this broader merge, but eventually we'd want this type of change as a separate patch.

File: plugins/handler/plugin_impl_handler.go [hunk 3]
```diff
@@ -71,11 +71,14 @@ func (p *Plugin) AfterInit() error {
 	p.Log.Info("AfterInit")
 
 	ticker := time.NewTicker(objectstore.ObjectStoreReadyInterval)
+	timeout := time.After(objectstore.ObjectStoreReadyTimeout)
```
Comment (mestery) [inline plugins/handler/plugin_impl_handler.go]: Same with this one, though this was also done and merged as part of #136 I believe.

File: plugins/nsmserver/nsmserver_plugin.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+	"github.com/ligato/cn-infra/logging"
+	"github.com/ligato/networkservicemesh/plugins/objectstore"
+	"github.com/ligato/networkservicemesh/utils/idempotent"
+)
+
+// Plugin watches K8s resources and causes all changes to be reflected in the ETCD
+// data store.
+type Plugin struct {
+	nsmClientEndpoints nsmClientEndpoints
+	Deps
```
Comment (mestery) [inline plugins/nsmserver/nsmserver_plugin.go]: I *think* you want Deps first here.

File: plugins/objectstore/networkservices.go [hunk 5]
```diff
@@ -48,6 +49,7 @@ func (n *networkServicesStore) Add(ns *netmesh.NetworkService) {
 		// Not in the store, adding it.
 		n.networkService[key] = ns
 	}
+	logrus.Infof("objectstore: NetworkService was added, %s/%s, total number of elements: %d", ns.Metadata.Namespace, ns.Metadata.Name, len(n.networkService))
```
Comment (mestery) [inline plugins/objectstore/networkservices.go]: I think we should use the Plugin's logger structure rather than logrus here.

File: scripts/integration-tests [hunk 6]
```diff
@@ -127,8 +126,11 @@ kubectl create -f conf/sample/networkservice-channel.yaml
 kubectl create -f conf/sample/networkservice-endpoint.yaml
 kubectl create -f conf/sample/networkservice.yaml
 kubectl create -f conf/sample/nsm-client.yaml
+#
+# Now let's wait for netmesh daemonset and nsm-cient pods to get into running state
+#
+wait_for_pods default
```
Comment (mestery) [inline scripts/integration-tests]: Just so I understand, you're starting the client after creating CRDs so the client can find the NetworkService's, right? It would be good to start the client pod up where the NSM daemonset is started here, but that may require the change to have the client either poll or be notified about NetworkService's being created. But this may be ok for now as it's expedient.

Comment (sbezverk) [reply scripts/integration-tests]: If you check the log produced by the test, you will see client pod in pending state, it is because we do not give it time to start. If nsm client pod encounters any issues or goes into a crashloop, without added wait_for_pod, we will never see it.

File: plugins/objectstore/networkservices.go (hunk 5 above)
Comment (sbezverk) [reply plugins/objectstore/networkservices.go]: right, good catch will change it.

File: plugins/nsmserver/nsmserver_plugin.go (hunk 4 above)
Comment (sbezverk) [reply plugins/nsmserver/nsmserver_plugin.go]: it does not change any logic, but to be consistent with other plugins, will change it.

File: cmd/nsm-init/nsm-init.go (hunk 1 above)
Comment (sbezverk) [reply cmd/nsm-init/nsm-init.go]: By nature, init container, if we are talking about init container, if one shot deal. It does not run at the same time with main application, it runs only once, right before the main app container and then it goes away. 
I think it is a good model, because even if it were running as a side car container and had detected that NSM got new NetworkService, it would still need to restart the whole POD so the application container would start using.

File: plugins/objectstore/networkservices.go (hunk 5 above)
Comment (sbezverk) [reply plugins/objectstore/networkservices.go]: Actually these are debug messages. (removed them) logging of event should be done by the plugin which called these methods.

File: cmd/nsm-init/nsm-init.go (hunk 1 above)
Comment (mestery) [reply cmd/nsm-init/nsm-init.go]: I think we may eventually want to evolve into something when we can dynamically modify the networking, but this makes sense for now. Thanks.

File: scripts/integration-tests (hunk 6 above)
Comment (mestery) [reply scripts/integration-tests]: Right, my point was, maybe move the start of the client pod to line 108, unless it is required that the NetworkServices exist before the client pod is started. Can you confirm?

Comment (mestery) [reply scripts/integration-tests]: OK, then I'd suggest moving to line 112 or so, so it looks like this:

```
kubectl create -f conf/sample/networkservice-daemonset.yaml
#
# Now let's wait for all pods to get into running state
#
wait_for_pods default

kubectl create -f conf/sample/nsm-client.yam

#
# Wait again for the client pod to be running
#
wait_for_pods default
```

Comment (sbezverk) [reply scripts/integration-tests]: Sounds good to me.
